### PR TITLE
Issue 800  product space getitem

### PR DESCRIPTION
**Pull request by adler-j**

### CODE REVIEW

File: odl/operator/pspace_ops.py [hunk 1]
```diff
@@ -354,9 +358,47 @@ def adjoint(self):
         adj_matrix = sp.sparse.coo_matrix((adjoint_ops, indices), shape)
         return ProductSpaceOperator(adj_matrix, self.range, self.domain)
 
+    def __getitem__(self, index):
+        """Get sub-operator by index.
+
+        Returns
+        -------
+        suboperator : `Operator` or ``None``
+            If there is an operator at [row, col], returns the operator,
```
Comment (adler-j) [inline odl/operator/pspace_ops.py]: > Unclear how [row, col] relate to index.

Will fix

>Also, BroadcastOperator and ReductionOperator should support 1d indexing, how about that?

They already do. Only one missing is DiagonalOperator

Comment (kohr-h) [inline odl/operator/pspace_ops.py]: > They already do.

Separate implementation? Otherwise it's a mystery to me when the matrix isn't squashed.

Comment (adler-j) [inline odl/operator/pspace_ops.py]: They aren't even `ProductSpaceOperators`, they only use them as an internal representation.

Comment (kohr-h) [inline odl/operator/pspace_ops.py]: Ah, mystery solved.

**Comment by adler-j** [comment]
Fixed comments, merge?

An obvious followup is to allow "single index access" of `ProductSpaceOperator` which would return a `ReductionOperator` or a `BroadcastOperator`. But we'll leave that for now since the uses cases are not really abundant.

**Comment by adler-j** [comment]
Bump

File: odl/operator/pspace_ops.py [hunk 2]
```diff
@@ -354,9 +358,51 @@ def adjoint(self):
...
+        Parameters
+        ----------
+        index : tuple of int
+            A pair of integers given as (row, col).
+
+        Returns
+        -------
+        suboperator : `Operator` or ``0``
+            If there is an operator at (row, col), the operator is returned,
+            otherwise returns ``0``.
```
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: ~~returns~~

File: odl/operator/pspace_ops.py [hunk 3]
```diff
@@ -354,9 +358,51 @@ def adjoint(self):
...
+        else:
+            return self.ops.data[int(linear_index)]
+
     def __repr__(self):
         """Return ``repr(self)``."""
-        return 'ProductSpaceOperator({!r})'.format(self.ops)
+        aslist = [[0] * self.domain.size for _ in range(self.range.size)]
+        for i, j, op in zip(self.ops.row, self.ops.col, self.ops.data):
+            aslist[i][j] = op
+        return 'ProductSpaceOperator({!r})'.format(aslist)
```
Comment (kohr-h) [inline odl/operator/pspace_ops.py]: `self.__class__.__name__`

**Comment by adler-j** [comment]
Merge after CI.

**Comment by kohr-h** [comment]
Do that, it's already mentioned in the release notes :-) #821

**Pull request merged**